Make cef.rb encoder compliant to spec.

fchouina opened the pull request:
The CEF specifications document stipulates that "When sending events, the CEF key name is the proper form to use" (this is the short name). The encoder was not compliant to this because it produced CEF messages with the "full name" instead of the "CEF key name". This modification introduces a reverse mapping for the fields that require it.

Thanks for contributing to Logstash! If you haven't already signed our CLA, here's a handy link: https://www.elastic.co/contributor-agreement/

jsvd commented:
do you mind adding a test here for this change? you can see similar specs for the encoding [here](https://github.com/logstash-plugins/logstash-codec-cef/blob/master/spec/codecs/cef_spec.rb#L12)

fchouina commented:
I am working on it.

fchouina commented:
I have completed the addition of a test case for validation of the encoded CEF key names.
Let me know if there is anything else.

fchouina commented:
Can you give me an idea of what release of the Elastic Stack we can expect this change to appear in?

colinsurprenant commented on lib/logstash/codecs/cef.rb:
```diff
@@ -75,6 +75,9 @@ class LogStash::Codecs::CEF < LogStash::Codecs::Base
   MAPPINGS = { "act" => "deviceAction", "app" => "applicationProtocol", "c6a1" => "deviceCustomIPv6Address1", "c6a1Label" => "deviceCustomIPv6Address1Label", "c6a2" => "deviceCustomIPv6Address2", "c6a2Label" => "deviceCustomIPv6Address2Label", "c6a3" => "deviceCustomIPv6Address3", "c6a3Label" => "deviceCustomIPv6Address3Label", "c6a4" => "deviceCustomIPv6Address4", "c6a4Label" => "deviceCustomIPv ...
 
+  # Reverse mapping of CEF full field names to CEF extensions field names for encoding into a CEF event for output.
+  REVERSE_MAPPINGS= { "deviceAction" => "act", "applicationProtocol" => "app",  "deviceCustomIPv6Address1" => "c6a1", "deviceCustomIPv6Address1Label" =>  "c6a1Label",  "deviceCustomIPv6Address2" => "c6a2",  "deviceCustomIPv6Address2Label" => "c6a2Label",  "deviceCustomIPv6Address3" =>  "c6a3",  "deviceCustomIPv6Address3Label" => "c6a3Label",  "deviceCustomIPv6Address4" => "c6a4",  "deviceCustomIP ...
```
if `REVERSE_MAPPINGS` is the exact inverse of `MAPPINGS` and `MAPPINGS` values are unique then you can use 
```rb
REVERSE_MAPPINGS = MAPPINGS.invert
```
instead of manually inverting it.

fchouina replied:
I just reviewed the fields and realized that there is 1 missing field in the original MAPPINGS:
"dvcmac" => deviceMacAddress

That is the only difference.
I believe it should be there.
I will make the changes and simply invert the MAPPINGS for REVERSE_MAPPINGS.

colinsurprenant commented on lib/logstash/codecs/cef.rb:
```diff
@@ -266,13 +269,16 @@ def get_value(fieldname, event)
 
     return nil if val.nil?
 
+    key = "#{sanitize_extension_key(fieldname)}"
```
since `sanitize_extension_key(fieldname)` returns a `String` there is no need to use string interpolation. You could just write:
```rb
key = sanitize_extension_key(fieldname)
```

fchouina replied:
Ok. Will adjust my pull request.

colinsurprenant commented on lib/logstash/codecs/cef.rb:
```diff
@@ -266,13 +269,16 @@ def get_value(fieldname, event)
     return nil if val.nil?
 
+    key = "#{sanitize_extension_key(fieldname)}"
+    key = "#{REVERSE_MAPPINGS[key] || key}"
```
same here, no need for string interpolation:
```rb
key = REVERSE_MAPPINGS[key] || key
```

fchouina replied:
Ok. Will adjust my pull request.

colinsurprenant commented on spec/codecs/cef_spec.rb:
```diff
@@ -209,6 +209,14 @@
+      event = LogStash::Event.new("deviceAction" => "foobar", "applicationProtocol" => "foobar",  "deviceCustomIPv6Address1" => "foobar", "deviceCustomIPv6Address1Label" =>  "foobar",  "deviceCustomIPv6Address2" => "foobar",  "deviceCustomIPv6Address2Label" => "foobar",  "deviceCustomIPv6Address3" =>  "foobar",  "deviceCustomIPv6Address3Label" => "foobar",  "deviceCustomIPv6Address4" => "foobar", ...
+      codec.encode(event)
+      expect(results.first).to match(/^CEF:0\|Elasticsearch\|Logstash\|1.0\|Logstash\|Logstash\|6\|act=foobar app=foobar c6a1=foobar c6a1Label=foobar c6a2=foobar c6a2Label=foobar c6a3=foobar c6a3Label=foobar c6a4=foobar c6a4Label=foobar cat=foobar cfp1=foobar cfp1Label=foobar cfp2=foobar cfp2Label=foobar cfp3=foobar cfp3Label=foobar cfp4=foobar cfp4Label=foobar cn1=foobar cn1Label=foobar cn2=foob ...
+    end
```
👍

colinsurprenant commented:
thanks for your contribution @fchouina - I left a few review comments for you to look at!

colinsurprenant commented on lib/logstash/codecs/cef.rb:
```diff
@@ -73,7 +73,10 @@ class LogStash::Codecs::CEF < LogStash::Codecs::Base
 
   # Translating and flattening the CEF extensions with known field names as documented in the Common Event Format whitepaper
-  MAPPINGS = { "act" => "deviceAction", "app" => "applicationProtocol", "c6a1" => "deviceCustomIPv6Address1", "c6a1Label" => "deviceCustomIPv6Address1Label", "c6a2" => "deviceCustomIPv6Address2", "c6a2Label" => "deviceCustomIPv6Address2Label", "c6a3" => "deviceCustomIPv6Address3", "c6a3Label" => "deviceCustomIPv6Address3Label", "c6a4" => "deviceCustomIPv6Address4", "c6a4Label" => "deviceCustomIPv ...
+  MAPPINGS = { "act" => "deviceAction", "app" => "applicationProtocol", "c6a1" => "deviceCustomIPv6Address1", "c6a1Label" => "deviceCustomIPv6Address1Label", "c6a2" => "deviceCustomIPv6Address2", "c6a2Label" => "deviceCustomIPv6Address2Label", "c6a3" => "deviceCustomIPv6Address3", "c6a3Label" => "deviceCustomIPv6Address3Label", "c6a4" => "deviceCustomIPv6Address4", "c6a4Label" => "deviceCustomIPv ...
```
I think you added  `"dvcmac" => "deviceMacAddress"` in the `MAPPINGS` right?

fchouina replied:
Yes as stated in my comment on your review comment about using MAPPINGS.invert for REVERSE_MAPPINGS.

colinsurprenant commented:
@fchouina The changes looks good. One thing I just realized is that with the introduction of the reverse mapping we are actually changing the event layout which means that anyone upgrading the codec will get new fields keys in their events. 

To avoid this BWC issue, we should introduce a new option, something like 
```
config :reverse_mapping, :validate => :boolean, :default => false
```
and default it to false so that we don't create any BWC issues. 
WDYT?

fchouina commented:
What about the addition of "dvcmac" =>"deviceMacAddress" to the MAPPINGS in that case?
If the field is present in the source event and the users have "strict" mapping templates, they will also be affected by this change.

colinsurprenant commented:
@fchouina good point. Let's revert that `"dvcmac" =>"deviceMacAddress" ` change here and let's create a separate PR just for that where we will be able to bump version number according to BWC.

fchouina commented:
Here are the new modifications I have just introduced:
01- I have sync'ed with master (5.0.5).
02- I have removed field "deviceMacAddress" and will make a separate pull request for this as suggested by Colin.
03- I have added the "reverse_mapping" configuration option as suggested by Colin, and set the default to "false".
04- I have added a conditional for this option when resolving the CEF field names for encoding to CEF.
05- I have adjusted the unit tests to test both with and without reverse mappings for the CEF field names.

I believe this addresses all the requirements I had to finalize this pull request.

colinsurprenant commented:
thanks @fchouina  - reviewing!

colinsurprenant commented:
LGTM - also tested locally. 
Good job @fchouina ! I'll go ahead and merge this and then submit a version bump PR so we can get that out soon.

colinsurprenant commented:
BTW @fchouina you can submit a PR agaist https://github.com/logstash-plugins/logstash-codec-cef/blob/master/CONTRIBUTORS to add your name if you want it added!

colinsurprenant commented:
version 5.0.6 published. thanks again @fchouina for your contribution.
